agent/session.go

```diff
@@ -54,6 +56,7 @@ func newSession(ctx context.Context, agent *Agent, delay time.Duration, sessionI
 		subscriptions: make(chan *api.SubscriptionMessage),
 		registered:    make(chan struct{}),
 		closed:        make(chan struct{}),
+		cancel:        sessionCancel,
```

Comment:
Maybe `(*session).start` should use this cancel function instead of forking the context again.